pkg/acs/calacs/acs2d/dodark.c

```diff
@@ -53,152 +52,166 @@
+     trlmessage(MsgText);
+  }
   
-	/* Get the dark image data. */
+  /* Initialze and get the dark image data */
+  initSingleGroup (&y);
+
   if (acs2d->pctecorr == PERFORM) {
-    openSingleGroupLine (acs2d->darkcte.name, extver, &y);
+      getSingleGroup (acs2d->darkcte.name, extver, &y);
   } else {
-    openSingleGroupLine (acs2d->dark.name, extver, &y);
+      getSingleGroup (acs2d->dark.name, extver, &y);
+  }
+
+  if (hstio_err()) {
+     freeSingleGroup (&y);
+     return (status = OPEN_FAILED);
+  }
+
+  /* Compare binning of science image and reference image; get same_size 
+     and high_res flags, and get info about binning and offset for 
+     use by bin2d.
+  */
+  if (FindBin (x, &y, &same_size, &rx, &ry, &x0, &y0)) {
+     freeSingleGroup (&y);
+     return (status);
   }
-	if (hstio_err())
-    return (status = OPEN_FAILED);
-  
-  
-	/* Compare binning of science image and reference image;
-   get same_size and high_res flags, and get info about
-   binning and offset for use by bin2d.
-   */
-	if (FindLine (x, &y, &same_size, &rx, &ry, &x0, &y0))
-    return (status);
   
   if (rx != 1 || ry != 1) {
-		sprintf(MsgText,"Reference image and input are not binned to the same pixel size!");
-		trlmessage(MsgText);
+     sprintf(MsgText,"Reference image and input are not binned to the same pixel size!");
+     trlmessage(MsgText);
+  }
+  if (acs2d->verbose){
+     sprintf(MsgText,"Image has an offset of %d,%d",x0,y0);
+     trlmessage(MsgText);
   }
-	if (acs2d->verbose){
-		sprintf(MsgText,"Image has an offset of %d,%d",x0,y0);
-		trlmessage(MsgText);
-	}
-  
-	mean = 0.0;
-  weight = 0.0;
-  
-	/* Bin the dark image down to the actual size of x. */
   
-	initSingleGroupLine (&z);
-	allocSingleGroupLine (&z, x->sci.data.nx);
-	for (i=0, j=y0; i < scilines; i++,j++) { 
+  /* Trim the dark image (y->z) down to the actual size of the science image (x). */
+  initSingleGroup (&z);
+  allocSingleGroup (&z, scicols, scirows, True);
+  if (hstio_err()) {
+     freeSingleGroup (&y);
+     freeSingleGroup (&z);
+     return (status = ALLOCATION_PROBLEM);
+  }
     
-    /* We are working with a sub-array and need to apply the
+  /* We are working with a sub-array and need to apply the
      proper section from the reference image to the science image.
-     */
-		if (acs2d->pctecorr == PERFORM) {
-      getSingleGroupLine (acs2d->darkcte.name, j, &y);
-    } else {
-      getSingleGroupLine (acs2d->dark.name, j, &y);
-    }
-    
-    /*
-     rx = 1;
-     */
-    update = NO;
-    
-    if (trim1d (&y, x0, y0, rx, avg, update, &z)) {
-			trlerror ("(darkcorr) size mismatch.");
-			return (status);
-    }
-    
-    multk1d(&z, darktime);
-    
-    AvgSciValLine (&z, acs2d->sdqflags, &dark, &wdark);
+     update = NO = do not update the header of the scratch image
+  */
+  update = NO;
+  if (trim2d (&y, x0, y0, rx, ry, update, &z)) {
+     trlerror ("(darkcorr) size mismatch.");
+     freeSingleGroup (&y);
```

Comment:
For future ref, I would have taken the opportunity to use the ``PtrRegister`` so as to remove the burden of knowing what is alive to be freed upon return.

Struct is in [include/hstcal_memory.h](https://github.com/spacetelescope/hstcal/blob/master/include/hstcal_memory.h) and the "methods" in [lib/hstcal_memory.c](https://github.com/spacetelescope/hstcal/blob/master/lib/hstcal_memory.c). [Here's an example](https://github.com/spacetelescope/hstcal/blob/master/include/hstcal_memory.h#L4-L32) - for ``SingleGroup``s you pass ``&freeSingleGroup`` as the free func.

Reply:
Excellent utility and implementing in routines I have modified.